File: gre_tunnel.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <linux/netlink.h>
#include <linux/rtnetlink.h>
#include <arpa/inet.h>
#include <net/if.h>
#include <linux/if_tunnel.h>

#define BUFFER_SIZE 8192
#define MAX_PAYLOAD 1024

struct gre_config {
    char tunnel_name[IFNAMSIZ];
    char local_ip[INET_ADDRSTRLEN];
    char remote_ip[INET_ADDRSTRLEN];
    char tunnel_local_ip[INET_ADDRSTRLEN];
    int tunnel_prefix;
    int ttl;
};

// Function to send netlink message
int send_netlink_msg(int sock, struct nlmsghdr *nlh) {
    struct sockaddr_nl dest_addr;
    struct iovec iov;
    struct msghdr msg;
    
    memset(&dest_addr, 0, sizeof(dest_addr));
    dest_addr.nl_family = AF_NETLINK;
    dest_addr.nl_pid = 0;    // Kernel
    dest_addr.nl_groups = 0;
    
    iov.iov_base = nlh;
    iov.iov_len = nlh->nlmsg_len;
    
    msg.msg_name = &dest_addr;
    msg.msg_namelen = sizeof(dest_addr);
    msg.msg_iov = &iov;
    msg.msg_iovlen = 1;
    msg.msg_control = NULL;
    msg.msg_controllen = 0;
    msg.msg_flags = 0;
    
    return sendmsg(sock, &msg, 0);
}

// Function to receive netlink response
int recv_netlink_msg(int sock) {
    char buffer[BUFFER_SIZE];
    struct nlmsghdr *nlh;
    int len;
    
    len = recv(sock, buffer, BUFFER_SIZE, 0);
    if (len < 0) {
        perror("recv");
        return -1;
    }
    
    nlh = (struct nlmsghdr *)buffer;
    if (nlh->nlmsg_type == NLMSG_ERROR) {
        struct nlmsgerr *err = (struct nlmsgerr *)NLMSG_DATA(nlh);
        if (err->error != 0) {
            fprintf(stderr, "Netlink error: %s\n", strerror(-err->error));
            return -1;
        }
    }
    
    return 0;
}

// Function to create GRE tunnel
int create_gre_tunnel(struct gre_config *config) {
    int sock;
    struct nlmsghdr *nlh;
    struct ifinfomsg *ifi;
    struct rtattr *linkinfo, *infodata;
    char buffer[BUFFER_SIZE];
    int seq = 1;
    
    printf("Creating GRE tunnel: %s\n", config->tunnel_name);
    printf("Local endpoint: %s\n", config->local_ip);
    printf("Remote endpoint: %s\n", config->remote_ip);
    printf("Tunnel IP: %s/%d\n", config->tunnel_local_ip, config->tunnel_prefix);
    
    // Create netlink socket
    sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
    if (sock < 0) {
        perror("socket");
        return -1;
    }
    
    // Prepare netlink message
    memset(buffer, 0, BUFFER_SIZE);
    nlh = (struct nlmsghdr *)buffer;
    nlh->nlmsg_len = NLMSG_LENGTH(sizeof(struct ifinfomsg));
    nlh->nlmsg_type = RTM_NEWLINK;
    nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_CREATE | NLM_F_EXCL;
    nlh->nlmsg_seq = seq++;
    nlh->nlmsg_pid = getpid();
    
    ifi = (struct ifinfomsg *)NLMSG_DATA(nlh);
    ifi->ifi_family = AF_UNSPEC;
    ifi->ifi_type = 0;
    ifi->ifi_index = 0;
    ifi->ifi_flags = 0;
    ifi->ifi_change = 0;
    
    // Add interface name
    struct rtattr *ifname = (struct rtattr *)(((char *)nlh) + NLMSG_ALIGN(nlh->nlmsg_len));
    ifname->rta_type = IFLA_IFNAME;
    ifname->rta_len = RTA_LENGTH(strlen(config->tunnel_name) + 1);
    strcpy(RTA_DATA(ifname), config->tunnel_name);
    nlh->nlmsg_len = NLMSG_ALIGN(nlh->nlmsg_len) + RTA_LENGTH(strlen(config->tunnel_name) + 1);
    
    // Add link info
    linkinfo = (struct rtattr *)(((char *)nlh) + NLMSG_ALIGN(nlh->nlmsg_len));
    linkinfo->rta_type = IFLA_LINKINFO;
    linkinfo->rta_len = RTA_LENGTH(0);
    
    // Add kind (gre)
    struct rtattr *kind = (struct rtattr *)(((char *)linkinfo) + RTA_LENGTH(0));
    kind->rta_type = IFLA_INFO_KIND;
    kind->rta_len = RTA_LENGTH(4);
    strcpy(RTA_DATA(kind), "gre");
    linkinfo->rta_len += RTA_LENGTH(4);
    
    // Add GRE-specific data
    infodata = (struct rtattr *)(((char *)linkinfo) + linkinfo->rta_len);
    infodata->rta_type = IFLA_INFO_DATA;
    infodata->rta_len = RTA_LENGTH(0);
    
    // Add local IP
    struct rtattr *local = (struct rtattr *)(((char *)infodata) + RTA_LENGTH(0));
    local->rta_type = IFLA_GRE_LOCAL;
    local->rta_len = RTA_LENGTH(sizeof(uint32_t));
    inet_pton(AF_INET, config->local_ip, RTA_DATA(local));
    infodata->rta_len += RTA_LENGTH(sizeof(uint32_t));
    
    // Add remote IP
    struct rtattr *remote = (struct rtattr *)(((char *)infodata) + infodata->rta_len);
    remote->rta_type = IFLA_GRE_REMOTE;
    remote->rta_len = RTA_LENGTH(sizeof(uint32_t));
    inet_pton(AF_INET, config->remote_ip, RTA_DATA(remote));
    infodata->rta_len += RTA_LENGTH(sizeof(uint32_t));
    
    // Add TTL
    struct rtattr *ttl = (struct rtattr *)(((char *)infodata) + infodata->rta_len);
    ttl->rta_type = IFLA_GRE_TTL;
    ttl->rta_len = RTA_LENGTH(sizeof(uint8_t));
    *(uint8_t *)RTA_DATA(ttl) = config->ttl;
    infodata->rta_len += RTA_LENGTH(sizeof(uint8_t));
    
    linkinfo->rta_len += infodata->rta_len;
    nlh->nlmsg_len = NLMSG_ALIGN(nlh->nlmsg_len) + linkinfo->rta_len;
    
    // Send message
    if (send_netlink_msg(sock, nlh) < 0) {
        perror("send_netlink_msg");
        close(sock);
        return -1;
    }
    
    // Receive response
    if (recv_netlink_msg(sock) < 0) {
        close(sock);
        return -1;
    }
    
    close(sock);
    printf("✓ GRE tunnel created successfully\n");
    return 0;
}

// Function to assign IP address to interface
int assign_ip_address(struct gre_config *config) {
    int sock;
    struct nlmsghdr *nlh;
    struct ifaddrmsg *ifa;
    char buffer[BUFFER_SIZE];
    int seq = 1;
    int if_index;
    
    printf("Assigning IP address to tunnel interface...\n");
    
    // Get interface index
    if_index = if_nametoindex(config->tunnel_name);
    if (if_index == 0) {
        perror("if_nametoindex");
        return -1;
    }
    
    // Create netlink socket
    sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
    if (sock < 0) {
        perror("socket");
        return -1;
    }
    
    // Prepare netlink message
    memset(buffer, 0, BUFFER_SIZE);
    nlh = (struct nlmsghdr *)buffer;
    nlh->nlmsg_len = NLMSG_LENGTH(sizeof(struct ifaddrmsg));
    nlh->nlmsg_type = RTM_NEWADDR;
    nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_CREATE | NLM_F_EXCL;
    nlh->nlmsg_seq = seq++;
    nlh->nlmsg_pid = getpid();
    
    ifa = (struct ifaddrmsg *)NLMSG_DATA(nlh);
    ifa->ifa_family = AF_INET;
    ifa->ifa_prefixlen = config->tunnel_prefix;
    ifa->ifa_flags = IFA_F_PERMANENT;
    ifa->ifa_scope = RT_SCOPE_UNIVERSE;
    ifa->ifa_index = if_index;
    
    // Add local address
    struct rtattr *local = (struct rtattr *)(((char *)nlh) + NLMSG_ALIGN(nlh->nlmsg_len));
    local->rta_type = IFA_LOCAL;
    local->rta_len = RTA_LENGTH(sizeof(uint32_t));
    inet_pton(AF_INET, config->tunnel_local_ip, RTA_DATA(local));
    nlh->nlmsg_len = NLMSG_ALIGN(nlh->nlmsg_len) + RTA_LENGTH(sizeof(uint32_t));
    
    // Add address
    struct rtattr *addr = (struct rtattr *)(((char *)nlh) + NLMSG_ALIGN(nlh->nlmsg_len));
    addr->rta_type = IFA_ADDRESS;
    addr->rta_len = RTA_LENGTH(sizeof(uint32_t));
    inet_pton(AF_INET, config->tunnel_local_ip, RTA_DATA(addr));
    nlh->nlmsg_len = NLMSG_ALIGN(nlh->nlmsg_len) + RTA_LENGTH(sizeof(uint32_t));
    
    // Send message
    if (send_netlink_msg(sock, nlh) < 0) {
        perror("send_netlink_msg");
        close(sock);
        return -1;
    }
    
    // Receive response
    if (recv_netlink_msg(sock) < 0) {
        close(sock);
        return -1;
    }
    
    close(sock);
    printf("✓ IP address assigned successfully\n");
    return 0;
}

// Function to bring interface up
int bring_interface_up(struct gre_config *config) {
    int sock;
    struct nlmsghdr *nlh;
    struct ifinfomsg *ifi;
    char buffer[BUFFER_SIZE];
    int seq = 1;
    int if_index;
    
    printf("Bringing interface up...\n");
    
    // Get interface index
    if_index = if_nametoindex(config->tunnel_name);
    if (if_index == 0) {
        perror("if_nametoindex");
        return -1;
    }
    
    // Create netlink socket
    sock = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
    if (sock < 0) {
        perror("socket");
        return -1;
    }
    
    // Prepare netlink message
    memset(buffer, 0, BUFFER_SIZE);
    nlh = (struct nlmsghdr *)buffer;
    nlh->nlmsg_len = NLMSG_LENGTH(sizeof(struct ifinfomsg));
    nlh->nlmsg_type = RTM_NEWLINK;
    nlh->nlmsg_flags = NLM_F_REQUEST;
    nlh->nlmsg_seq = seq++;
    nlh->nlmsg_pid = getpid();
    
    ifi = (struct ifinfomsg *)NLMSG_DATA(nlh);
    ifi->ifi_family = AF_UNSPEC;
    ifi->ifi_type = 0;
    ifi->ifi_index = if_index;
    ifi->ifi_flags = IFF_UP;
    ifi->ifi_change = IFF_UP;
    
    // Send message
    if (send_netlink_msg(sock, nlh) < 0) {
        perror("send_netlink_msg");
        close(sock);
        return -1;
    }
    
    // Receive response
    if (recv_netlink_msg(sock) < 0) {
        close(sock);
        return -1;
    }
    
    close(sock);
    printf("✓ Interface brought up successfully\n");
    return 0;
}

int main(int argc, char *argv[]) {
    struct gre_config config;
    
    // Default configuration with 10.x.x.x network
    strcpy(config.tunnel_name, "gre1");
    strcpy(config.local_ip, "10.1.1.10");      // Local endpoint IP
    strcpy(config.remote_ip, "10.1.1.20");     // Remote endpoint IP
    strcpy(config.tunnel_local_ip, "10.10.1.1"); // Tunnel local IP
    config.tunnel_prefix = 24;
    config.ttl = 64;
    
    // Parse command line arguments if provided
    if (argc >= 3) {
        strcpy(config.local_ip, argv[1]);
        strcpy(config.remote_ip, argv[2]);
    }
    if (argc >= 4) {
        strcpy(config.tunnel_local_ip, argv[3]);
    }
    
    printf("=== GRE Tunnel Creator ===\n");
    printf("This program creates a GRE tunnel programmatically using netlink sockets\n\n");
    
    // Check if running as root
    if (geteuid() != 0) {
        fprintf(stderr, "This program must be run as root\n");
        return 1;
    }
    
    // Create GRE tunnel
    if (create_gre_tunnel(&config) < 0) {
        fprintf(stderr, "Failed to create GRE tunnel\n");
        return 1;
    }
    
    // Assign IP address
    if (assign_ip_address(&config) < 0) {
        fprintf(stderr, "Failed to assign IP address\n");
        return 1;
    }
    
    // Bring interface up
    if (bring_interface_up(&config) < 0) {
        fprintf(stderr, "Failed to bring interface up\n");
        return 1;
    }
    
    printf("\n=== GRE Tunnel Setup Complete ===\n");
    printf("Tunnel name: %s\n", config.tunnel_name);
    printf("Local endpoint: %s\n", config.local_ip);
    printf("Remote endpoint: %s\n", config.remote_ip);
    printf("Tunnel IP: %s/%d\n", config.tunnel_local_ip, config.tunnel_prefix);
    printf("\nUsage: %s [local_ip] [remote_ip] [tunnel_ip]\n", argv[0]);
    
    return 0;
}

File: gre_tunnel.py
#!/usr/bin/env python3
"""
GRE Tunnel Creator - Python Implementation
This program creates GRE tunnels programmatically using pyroute2 library.
Uses 10.x.x.x network range to avoid conflicts with common LAN setups.
"""

import sys
import os
import socket
import struct
from contextlib import contextmanager

try:
    from pyroute2 import IPRoute
    PYROUTE2_AVAILABLE = True
except ImportError:
    PYROUTE2_AVAILABLE = False

class GREConfig:
    def __init__(self):
        self.tunnel_name = "gre1"
        self.local_ip = "10.1.1.10"         # Local endpoint IP
        self.remote_ip = "10.1.1.20"        # Remote endpoint IP  
        self.tunnel_local_ip = "10.10.1.1"  # Tunnel local IP
        self.tunnel_prefix = 24
        self.ttl = 64

class GRETunnelCreator:
    def __init__(self, config):
        self.config = config
        self.ip = None
        
    def __enter__(self):
        if not PYROUTE2_AVAILABLE:
            raise ImportError("pyroute2 library not found. Install with: pip install pyroute2")
        
        self.ip = IPRoute()
        return self
    
    def __exit__(self, exc_type, exc_val, exc_tb):
        if self.ip:
            self.ip.close()
    
    def create_tunnel(self):
        """Create GRE tunnel interface"""
        print(f"Creating GRE tunnel: {self.config.tunnel_name}")
        print(f"Local endpoint: {self.config.local_ip}")
        print(f"Remote endpoint: {self.config.remote_ip}")
        print(f"Tunnel IP: {self.config.tunnel_local_ip}/{self.config.tunnel_prefix}")
        
        try:
            # Create GRE tunnel
            self.ip.link('add', 
                        ifname=self.config.tunnel_name,
                        kind='gre',
                        gre_local=self.config.local_ip,
                        gre_remote=self.config.remote_ip,
                        gre_ttl=self.config.ttl)
            
            print("✓ GRE tunnel created successfully")
            return True
            
        except Exception as e:
            print(f"✗ Failed to create GRE tunnel: {e}")
            return False
    
    def assign_ip_address(self):
        """Assign IP address to tunnel interface"""
        print("Assigning IP address to tunnel interface...")
        
        try:
            # Get interface index
            idx = self.ip.link_lookup(ifname=self.config.tunnel_name)[0]
            
            # Add IP address
            self.ip.addr('add', 
                        index=idx,
                        address=self.config.tunnel_local_ip,
                        prefixlen=self.config.tunnel_prefix)
            
            print("✓ IP address assigned successfully")
            return True
            
        except Exception as e:
            print(f"✗ Failed to assign IP address: {e}")
            return False
    
    def bring_interface_up(self):
        """Bring tunnel interface up"""
        print("Bringing interface up...")
        
        try:
            # Get interface index
            idx = self.ip.link_lookup(ifname=self.config.tunnel_name)[0]
            
            # Bring interface up
            self.ip.link('set', index=idx, state='up')
            
            print("✓ Interface brought up successfully")
            return True
            
        except Exception as e:
            print(f"✗ Failed to bring interface up: {e}")
            return False
    
    def show_tunnel_info(self):
        """Display tunnel information"""
        print("\n=== Tunnel Information ===")
        
        try:
            # Get interface index
            idx = self.ip.link_lookup(ifname=self.config.tunnel_name)[0]
            
            # Get interface info
            links = self.ip.get_links(idx)
            if links:
                link = links[0]
                print(f"Interface: {link.get_attr('IFLA_IFNAME')}")
                print(f"State: {'UP' if link['state'] == 'up' else 'DOWN'}")
                print(f"MTU: {link.get_attr('IFLA_MTU')}")
            
            # Get addresses
            addrs = self.ip.get_addr(index=idx)
            if addrs:
                for addr in addrs:
                    if addr['family'] == socket.AF_INET:
                        ip_addr = addr.get_attr('IFA_ADDRESS')
                        prefix = addr['prefixlen']
                        print(f"IP Address: {ip_addr}/{prefix}")
            
            return True
            
        except Exception as e:
            print(f"✗ Failed to get tunnel info: {e}")
            return False
    
    def remove_tunnel(self):
        """Remove GRE tunnel"""
        print(f"Removing GRE tunnel: {self.config.tunnel_name}")
        
        try:
            # Get interface index
            idx = self.ip.link_lookup(ifname=self.config.tunnel_name)[0]
            
            # Remove tunnel
            self.ip.link('del', index=idx)
            
            print("✓ GRE tunnel removed successfully")
            return True
            
        except Exception as e:
            print(f"✗ Failed to remove tunnel: {e}")
            return False
    
    def test_connectivity(self):
        """Test tunnel connectivity"""
        print(f"Testing connectivity to remote tunnel endpoint...")
        
        # Extract IP without prefix
        remote_tunnel_ip = self.config.tunnel_local_ip.replace('1', '2')  # Simple assumption
        
        try:
            # Simple ping test using os.system
            result = os.system(f"ping -c 3 -W 2 {remote_tunnel_ip} > /dev/null 2>&1")
            if result == 0:
                print("✓ Tunnel connectivity test passed")
            else:
                print("⚠ Cannot reach remote endpoint (expected if remote side not configured)")
            return True
            
        except Exception as e:
            print(f"✗ Connectivity test failed: {e}")
            return False

def install_dependencies():
    """Install required dependencies"""
    print("Installing required dependencies...")
    try:
        os.system("pip3 install pyroute2")
        print("✓ Dependencies installed successfully")
        return True
    except Exception as e:
        print(f"✗ Failed to install dependencies: {e}")
        return False

def main():
    if len(sys.argv) > 1 and sys.argv[1] == "install-deps":
        install_dependencies()
        return
    
    # Check if running as root
    if os.geteuid() != 0:
        print("Error: This program must be run as root")
        print("Use: sudo python3 gre_tunnel.py")
        return 1
    
    # Check if pyroute2 is available
    if not PYROUTE2_AVAILABLE:
        print("Error: pyroute2 library not found")
        print("Install with: python3 gre_tunnel.py install-deps")
        print("Or manually: pip3 install pyroute2")
        return 1
    
    # Create configuration
    config = GREConfig()
    
    # Parse command line arguments
    if len(sys.argv) >= 3:
        config.local_ip = sys.argv[1]
        config.remote_ip = sys.argv[2]
    
    if len(sys.argv) >= 4:
        config.tunnel_local_ip = sys.argv[3]
    
    print("=== GRE Tunnel Creator (Python) ===")
    print("This program creates GRE tunnels programmatically using pyroute2\n")
    
    # Handle different commands
    command = sys.argv[-1] if len(sys.argv) > 1 else "create"
    
    try:
        with GRETunnelCreator(config) as tunnel_creator:
            if command == "remove":
                tunnel_creator.remove_tunnel()
            elif command == "status":
                tunnel_creator.show_tunnel_info()
            else:  # create (default)
                # Create tunnel
                if not tunnel_creator.create_tunnel():
                    return 1
                
                # Assign IP address
                if not tunnel_creator.assign_ip_address():
                    return 1
                
                # Bring interface up
                if not tunnel_creator.bring_interface_up():
                    return 1
                
                # Show tunnel info
                tunnel_creator.show_tunnel_info()
                
                # Test connectivity
                tunnel_creator.test_connectivity()
                
                print("\n=== GRE Tunnel Setup Complete ===")
                print(f"Tunnel name: {config.tunnel_name}")
                print(f"Local endpoint: {config.local_ip}")
                print(f"Remote endpoint: {config.remote_ip}")
                print(f"Tunnel IP: {config.tunnel_local_ip}/{config.tunnel_prefix}")
                print(f"\nUsage: python3 {sys.argv[0]} [local_ip] [remote_ip] [tunnel_ip]")
                print(f"Remove: python3 {sys.argv[0]} remove")
                print(f"Status: python3 {sys.argv[0]} status")
    
    except Exception as e:
        print(f"Error: {e}")
        return 1
    
    return 0

if __name__ == "__main__":
    sys.exit(main())